### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by mikeza** [tensorflow/magenta-studio#13]
Hi! I just found this project and am curious what the barriers are to building on linux?

Happy to help contribute to make this work.

Thanks!
Michael

**Comment by UnknownLocation** [comment]
Also I don't understand why standalone version for Windows still isn't available.

**Comment by adarob** [comment]
Re: Linux, this should be easy but I haven't looked into it. @teropa did you have a look at this?
Re: Windows, we have limited limited cycles right now, and we haven't prioritized it yet. I am looking into it this week, but there are some difficulties around dealing with DLLs etc. Also none if us use Windows so there is a learning curve :)

**Comment by mikeza** [comment]
So I actually just cloned the code and built on linux successfully.

The following was necessary:
1. I had to install some dependencies manually
2. I added linux to the output options and nameToPlatform in build.js
3. I printed the temporary build directory since linux binaries are extensionless and not found by the regex, also I stopped the temp directory from being deleted so I could test the build

It looks like the app works too, I just opened a couple and then tested generate since I didn't need midi files to train it with.

**Comment by adarob** [comment]
@mikeza would you might contributing back changes you made to get this working?

**Comment by mikeza** [comment]
Yeah I'd be interested in working on this, let me figure out what's up with the dependency installation.

**Comment by adarob** [comment]
I just tried building for linux on my mac, and the resulting build seems to work. However, I am not sure how to package it with an installer since the executable seems to need to live in the same folder as all of the resources.

**Comment by adarob** [comment]
With #16, you should be able to build the standalones with `npm run --output=linux-standalone`. I'll leave this open until I have a chance to add actual packages.

**Comment by mikeza** [comment]
> I just tried building for linux on my mac, and the resulting build seems to work on a linux machine. However, I am not sure how to package it with an installer since the executable seems to need to live in the same folder as all of the resources.

Maybe consider referring to https://github.com/electron-userland/electron-packager for best practice packaging the resources and executable (or could just use the project).

**Comment by tank-trax** [comment]
I have successfully built this in February on Debian Stretch following instructions that a member of the Facebook Linux Musicians posted, as follows:

1. git clone https://github.com/tensorflow/magenta-studio
2. follow the install instructions to get the npm node modules installed
3. In ./electron/build.js add `linux: 'linux' `to the hash nameToPlatform and comment out 
```
windows : 'win32',
macOS : 'darwin'
```
4. in moveFiles, comment out the await fs.remove(buildDir)
5. The temp build directory is in the /tmp/<whatever mktemp allocated>, and the electron apps for each working application is there separately.

**Comment by bkmgit** [comment]
Is it worth also making the Linux build downloadable, in addition to the sourcecode? [Snap](https://snapcraft.io/) might be one way.  [Flatpak](https://flatpak.org/) and [Appimages](https://appimage.org/) are possible other ways.

**Comment by han-so1omon** [comment]
For me on Archlinux, the app does not function correctly and renders as a gray box. I built using `yarn install && yarn build linux-standalone`. I also had to manually install gconf dependency.

Sidenote: User `flying-sheep` comments on the AUR gconf package that gconf typically is no longer a dependency of Electron apps. Is that the case here as well?

**Comment by Joseppi83** [comment]
Hello, fellow Linux users.  

To get this to properly install with Linux, you need to install nvm with the following:
`curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.0/install.sh | bash
`
Then, change your node version to 14 with the following:
`nvm install --14`

After that, you should be able to clone the repo, cd into the magenta-studio-master directory, and enter this terminal prompt.

`npm install`

Once it installs, go back to the README, and follow the instructions following `npm install`

**Comment by tank-trax** [comment]
@Joseppi83 

I think you mean `npm install --14`

following your instructions the apps do build, execute and are functional, but they still produce the same errors as below and in https://github.com/magenta/magenta-studio/issues/46#

```
ERROR:buffer_manager.cc(488)] [.DisplayCompositor]GL ERROR :GL_INVALID_OPERATION : glBufferData: <- error from previous GL command
```

![image](https://user-images.githubusercontent.com/41337033/220731488-01074059-8398-4c18-bc9c-58bd5e11994c.png)

**Comment by Joseppi83** [comment]
@tank-trax 

No, sir.  I meant `nvm install --14`, as in node version manager.  The previous command I printed installs nvm.  This allows you to quickly change and run different versions of node.

I get the exact same error you have in ubuntu.  However, I cannot add MIDI files to the standalone apps, but the UI appears.

**Comment by tank-trax** [comment]
@Joseppi83 OK had to close terminal it seems

however
```
nvm install --14
grep: unrecognized option '--14'
```

**Comment by Joseppi83** [comment]
@tank-trax 

Hmmm.  No idea.  According to the nvm docs [,](https://github.com/nvm-sh/nvm) you typed the correct command.  And for some reason, a flag to install that does not exist in the docs worked for me with Ubuntu.

**Comment by tank-trax** [comment]
@Joseppi83 

the command example they use is 

`nvm install 12`

**Comment by Joseppi83** [comment]
@tank-trax 

Seems we're both really eager to get this up and running on our Linux distros.  I have sooooo many musical ideas in my head.  And these standalone apps would help me materialize them into something tangible.

**Comment by tank-trax** [comment]
@Joseppi83 I have a working copy using the older commit and have posted functional binaries in my fork... and yes for my purposes Magenta-Studio is the perfect percussionist

**Comment by Joseppi83** [comment]
@tank-trax 

Yeah, I saw your GitHub post last night, when I was trying to get the standalones to work.  What you compiled was identical to what I compiled, but I produced the same result.  The UI appears with the JavaScript error, but I am unable to add MIDI files to the standalones.

**Comment by tank-trax** [comment]
@Joseppi83 because I am always on the lookout for the newest version of any app I tried this but it failed again. as well the apps are not working after testing

my builds work... I have tested them thoroughly. If you want to build them yourself follow the steps below. They should work with the NPM you have set up.. am trying now just to make sure

```
git clone https://github.com/magenta/magenta-studio.git
cd magenta-studio
git checkout c76f20c
npm install
npm run build linux-standalone
```

**Comment by Joseppi83** [comment]
@tank-trax 

I tried the above steps multiple times last night, but with no success.

Just to save some time, I have the stock SSD with Windows for my current laptop in storage.  I might just reinstall it, set up a dual boot with Linux, use Windows to produce the functional standalones, then transfer it to my free Linux DAW on the Linux side of the dual-boot.

I appreciate the effort though, sir.  Cheers!

**Comment by tank-trax** [comment]
sorry to hear that

have you tried my binaries? if they don't work your system may be missing deps

try any one of the executables with ldd

for example `ldd Continue`

you may want to try and clear out the folders like I did

~/.node-gyp/
~/.npm/
~/.electron/

I also cleared ~/.cache using bleachbit

 I have successfully built these packages on Mint 20.3 and Debian 11 without a hitch by literally copy and pasting those five lines into a terminal and pressing **Enter**

**Comment by Joseppi83** [comment]
@tank-trax 

Found out the problem.  I was lacking python2.  That was the only thing keeping me from successfully running the standalones.  Oy!  It's usually the smallest overlooked detail that gets me.

I really only wanted the drumify to see if it would create a cooler drum effect than I produced on LMMS.  It was really impressive, but it didn't quite capture the feeling I wanted to create.  

In the end, I chose my drum creation.

But I will definitely use these standalones in the future!  I'm constantly vomitting lyrics and acoustic guitar melodies and progressions.  I can use these apps to stimulate and inspire the other components to the whole.
